src/librustc_middle/ty/error.rs

```diff
@@ -513,4 +513,344 @@ impl Trait for X {
+        }
+        if let (Some(hir_id), false) = (self.hir().as_local_hir_id(body_owner_def_id), suggested) {
+            // When `body_owner` is an `impl` or `trait` item, look in its associated types for
+            // `expected` and point at it.
+            let parent_id = self.hir().get_parent_item(hir_id);
+            let item = self.hir().find(parent_id);
+            debug!("expected_projection parent item {:?}", item);
+            match item {
+                Some(hir::Node::Item(hir::Item {
+                    kind: hir::ItemKind::Trait(.., items), ..
+                })) => {
+                    // FIXME: account for `#![feature(specialization)]`
+                    for item in &items[..] {
+                        match item.kind {
+                            hir::AssocItemKind::Type | hir::AssocItemKind::OpaqueTy => {
+                                if self.type_of(self.hir().local_def_id(item.id.hir_id))
+                                    == values.found
+                                {
+                                    if let hir::Defaultness::Default { has_value: true } =
+                                        item.defaultness
+                                    {
+                                        db.span_label(
+                                            item.span,
+                                            "associated type defaults can't be assumed inside the \
+                                                trait defining them",
+                                        );
+                                    } else {
+                                        db.span_label(item.span, "expected this associated type");
+                                    }
+                                    suggested = true;
+                                }
+                            }
+                            _ => {}
+                        }
+                    }
+                }
+                Some(hir::Node::Item(hir::Item {
+                    kind: hir::ItemKind::Impl { items, .. },
+                    ..
+                })) => {
+                    for item in &items[..] {
+                        match item.kind {
+                            hir::AssocItemKind::Type | hir::AssocItemKind::OpaqueTy => {
+                                if self.type_of(self.hir().local_def_id(item.id.hir_id))
+                                    == values.found
+                                {
+                                    db.span_label(item.span, "expected this associated type");
+                                    suggested = true;
+                                }
+                            }
+                            _ => {}
+                        }
+                    }
+                }
+                _ => {}
+            }
+        }
+        if let ty::Opaque(def_id, _) = proj_ty.self_ty().kind {
```

Comment:
why is this done without a condition on `suggested`?

Reply:
The logic for the suggestions was already mutually exclusive. Now with the change to early returns this shouldn't be ambiguous anymore.